src/Reports/Junit.php

```diff
@@ -120,10 +136,34 @@ public function generate(
         }
 
         $failures = ($totalErrors + $totalWarnings);
-        echo '<?xml version="1.0" encoding="UTF-8"?>'.PHP_EOL;
-        echo '<testsuites name="PHP_CodeSniffer '.Config::VERSION.'" errors="0" tests="'.$tests.'" failures="'.$failures.'">'.PHP_EOL;
-        echo $cachedData;
-        echo '</testsuites>'.PHP_EOL;
+
+        $dom = new \DOMDocument();
```

Comment:
Please be aware that the DOM extension [can be disabled](https://www.php.net/manual/en/dom.installation.php)ay not be available and is currently not a requirement for PHPCS.

The [XMLWriter](https://www.php.net/manual/en/book.xmlwriter.php) extension and the [SimpleXML](https://www.php.net/manual/en/book.simplexml.php) extension _are_ requirements, so please consider if it's possible to use those instead.

Refs:
* https://github.com/squizlabs/PHP_CodeSniffer/blob/ce62dee92da3b038a399a99353fc055b39d2853e/composer.json#L27-L31
* https://github.com/squizlabs/PHP_CodeSniffer/blob/ce62dee92da3b038a399a99353fc055b39d2853e/src/Runner.php#L246-L257

Reply:
> Please be aware that the DOM extension [can be disabled](https://www.php.net/manual/en/dom.installation.php), so may not be available and is currently not a requirement for PHPCS.

I see. This is my first time writing PHP, so I assumed everything that's on PHP.net is shipped as default.

> The [XMLWriter](https://www.php.net/manual/en/book.xmlwriter.php) extension and the [SimpleXML](https://www.php.net/manual/en/book.simplexml.php) extension _are_ requirements, so please consider if it's possible to use those instead.

That depends on whether you consider consistent indentation important or not. 
- If indentation doesn't matter, then all the changes in `generate()` can be dropped.
- Otherwise, I looked through the docs for XMLWriter and SimpleXML several times but did not find ways to work with the partial XML that's emitted by `generateFileReport()`. I couldn't figure out if implementing the echo **within** this function is essential or if I could just create a class member of `Junit` holding the XMLWriter, append content to that during calls of `generateFileReport()` and echo all of it at the end of `generate()`. That would also ensure consistent indentation and likely not require DOM.

Reply:
> This is my first time writing PHP

Oh wow, respect for taking something like this on straight off the bat in that case :+1: 

> if I could just create a class member of Junit holding the XMLWriter, append content to that during calls of generateFileReport() and echo all of it at the end of generate()

If I remember correctly, the `echo` in `generateFileReport()` is needed with how things are set up in PHPCS now, but this may be a legacy from PHP 4 when PHP didn't have a proper object model. (this project is quite old).
The downside of saving it to a class property instead of to a file, would be the memory this would consume, which could be extensive on a large project with lots of violations.

I suppose you could just save the minimum information you need from each file in non-XML format, like is done for the `Summary` report, and then use the `generate()` method to parse the saved information and format it the way you want it to be.
Will be less efficient as the data is processed twice, but could possibly get you closer to the intended result.